coding/huffman.hpp

```diff
@@ -4,18 +4,74 @@
 #include "coding/varint.hpp"
 
 #include "base/assert.hpp"
+#include "base/checked_cast.hpp"
 #include "base/string_utils.hpp"
 
 #include "std/algorithm.hpp"
+#include "std/iterator.hpp"
 #include "std/map.hpp"
 #include "std/queue.hpp"
+#include "std/type_traits.hpp"
+#include "std/unique_ptr.hpp"
 #include "std/vector.hpp"
 
 namespace coding
 {
 class HuffmanCoder
 {
 public:
+  class Freqs
+  {
+  public:
+    using Table = map<uint32_t, uint32_t>;
+
+    Freqs() = default;
+
+    template <typename... Args>
+    Freqs(Args const &... args)
+    {
+      Add(args...);
+    }
+
+    void Add(strings::UniString const & s) { Add(s.begin(), s.end()); }
+
+    void Add(string const & s) { Add(s.begin(), s.end()); }
+
+    template <typename T>
+    void Add(T const * begin, T const * const end)
+    {
+      static_assert(sizeof(*begin) <= 4, "");
+      AddImpl(begin, end);
+    }
+
+    template <typename It>
+    void Add(It begin, It const end)
+    {
+      static_assert(sizeof(*begin) <= 4, "");
+      static_assert(is_integral<typename It::value_type>::value, "");
```

Comment:
Unfortunately, it's not possible, because sometimes a pair of pointers is passed to Add, and pointers don't have value_type.